src/experimental/patterns/ec2-app.ts

```diff
@@ -0,0 +1,307 @@
+    const { applicationPort } = props;
+    const { region, stackId } = scope;
+
+    super(scope, {
+      ...props,
+      updatePolicy: UpdatePolicy.rollingUpdate({
+        maxBatchSize: maximumInstances,
+        minInstancesInService: minimumInstances,
+        minSuccessPercentage: 100,
+        waitOnResourceSignals: true,
+
+        /*
+        If a scale-in event fires during an `AutoScalingRollingUpdate` operation, the update could fail and rollback.
+        For this reason, we suspend the `AlarmNotification` process, else availability of a service cannot be guaranteed.
+        Consequently, services cannot scale-out during deployments.
+        If AWS ever supports suspending scale-out and scale-in independently, we should allow scale-out.
+         */
+        suspendProcesses: [ScalingProcess.ALARM_NOTIFICATION],
+      }),
+    });
+
+    const { autoScalingGroup, targetGroup } = this;
+    const { userData, role } = autoScalingGroup;
+    const cfnAutoScalingGroup = autoScalingGroup.node.defaultChild as CfnAutoScalingGroup;
+
+    cfnAutoScalingGroup.desiredCapacity = minimumInstances.toString();
+
+    /*
+    Enable all ASG metrics for improved observability.
+    Specifying `granularity` without specifying `metrics` results in all metrics being enabled.
+    See https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-autoscaling-autoscalinggroup-metricscollection.html
+     */
+    cfnAutoScalingGroup.metricsCollection = [
+      {
+        granularity: "1Minute",
+      },
+    ];
+
+    // TODO are these sensible values?
+    const signalTimeoutSeconds = Math.max(
+      targetGroup.healthCheck.timeout?.toSeconds() ?? 0,
+      cfnAutoScalingGroup.healthCheckGracePeriod ?? 0,
+      Duration.minutes(5).toSeconds(),
+    );
+
+    const currentRollingUpdate = cfnAutoScalingGroup.cfnOptions.updatePolicy?.autoScalingRollingUpdate;
+
+    cfnAutoScalingGroup.cfnOptions.updatePolicy = {
+      autoScalingRollingUpdate: {
+        ...currentRollingUpdate,
+        pauseTime: Duration.seconds(signalTimeoutSeconds).toIsoString(),
+      },
+    };
+
+    cfnAutoScalingGroup.cfnOptions.creationPolicy = {
+      autoScalingCreationPolicy: {
+        minSuccessfulInstancesPercent: 100,
+      },
+      resourceSignal: {
+        count: minimumInstances,
+        timeout: Duration.seconds(signalTimeoutSeconds).toIsoString(),
+      },
+    };
+
+    AsgRollingUpdatePolicy.getInstance(scope).attachToRole(role);
+
+    /*
+    `aws` is available via AMIgo baked AMIs.
+    See https://github.com/guardian/amigo/tree/main/roles/aws-tools.
+     */
+    userData.addCommands(
+      `# ${GuEc2AppExperimental.name} UserData Start`,
+      `
+      TOKEN=$(curl -X PUT "http://169.254.169.254/latest/api/token" -H "X-aws-ec2-metadata-token-ttl-seconds: 21600")
+      INSTANCE_ID=$(curl -H "X-aws-ec2-metadata-token: $TOKEN" "http://169.254.169.254/latest/meta-data/instance-id")
```

Comment:
Curiously the [`ec2metadata` command](https://github.com/canonical/cloud-utils/blob/main/bin/ec2metadata) was failing with a 401 in AMIable CODE in deployTools account. IIUC a 401 is returned when an IMDSv2 token is not provided. I'm not sure how this is possible, as [`ec2metadata` only works with IMDSv2](https://github.com/canonical/cloud-utils/blob/49e5dd7849ee3c662f3db35e857148d02e72694b/bin/ec2metadata#L132).

Using the [AWS way](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html) of getting the instance ID seems to be more reliable. Not sure why. I can only conclude it's something with the AWS account, as [guardian/cdk-playground](https://github.com/guardian/cdk-playground/pull/522) and [guardian/testing-asg-rolling-update](https://github.com/guardian/testing-asg-rolling-update/pull/15) are fine in the developerPlayground account.

<details><summary>stack trace</summary>
<p>

```console
root@ip-10-248-51-213:/var/lib/cloud/instance# ec2metadata --instance-id
Traceback (most recent call last):
  File "/usr/bin/ec2metadata", line 249, in <module>
    main()
  File "/usr/bin/ec2metadata", line 245, in main
    display(metaopts, burl, prefix)
  File "/usr/bin/ec2metadata", line 192, in display
    value = m.get(metaopt)
  File "/usr/bin/ec2metadata", line 177, in get
    return self._get('meta-data/' + metaopt)
  File "/usr/bin/ec2metadata", line 137, in _get
    resp = urllib_request.urlopen(urllib_request.Request(url))
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 401: Unautho
```

</p>
</details>